pkg/simple/client/logging/elasticsearch/elasticsearch.go

```diff
@@ -34,76 +40,97 @@ type client interface {
 }
 
 func NewElasticsearch(options *Options) (*Elasticsearch, error) {
-	var version, index string
-	es := &Elasticsearch{}
-
-	if options.Version == "" {
-		var err error
-		version, err = detectVersionMajor(options.Host)
-		if err != nil {
-			return nil, err
-		}
-	} else {
-		version = options.Version
-	}
-
-	if options.IndexPrefix != "" {
-		index = options.IndexPrefix
-	} else {
-		index = "logstash"
+	var err error
+	es := &Elasticsearch{
+		host:    options.Host,
+		version: options.Version,
+		index:   options.IndexPrefix,
 	}
 
-	switch version {
+	switch es.version {
 	case ElasticV5:
-		es.c = v5.New(options.Host, index)
+		es.c, err = v5.New(es.host, es.index)
 	case ElasticV6:
-		es.c = v6.New(options.Host, index)
+		es.c, err = v6.New(es.host, es.index)
 	case ElasticV7:
-		es.c = v7.New(options.Host, index)
+		es.c, err = v7.New(es.host, es.index)
+	case "":
+		es.c = nil
 	default:
-		return nil, fmt.Errorf("unsupported elasticsearch version %s", version)
+		return nil, fmt.Errorf("unsupported elasticsearch version %s", es.version)
 	}
 
-	return es, nil
+	return es, err
 }
 
-func (es *Elasticsearch) ES() *client {
-	return &es.c
+func (es *Elasticsearch) getClient() client {
+	es.mux.RLock()
+	defer es.mux.RUnlock()
+
+	return es.c
 }
 
-func detectVersionMajor(host string) (string, error) {
+func (es *Elasticsearch) initClient() error {
 	// Info APIs are backward compatible with versions of v5.x, v6.x and v7.x
-	es := v6.New(host, "")
-	res, err := es.Client.Info(
-		es.Client.Info.WithContext(context.Background()),
+	esv6, err := v6.New(es.host, "")
+	if err != nil {
+		return err
+	}
+
+	res, err := esv6.Client.Info(
+		esv6.Client.Info.WithContext(context.Background()),
 	)
 	if err != nil {
-		return "", err
+		return err
 	}
 	defer res.Body.Close()
 
 	var b map[string]interface{}
 	if err = jsoniter.NewDecoder(res.Body).Decode(&b); err != nil {
-		return "", err
+		return err
 	}
 	if res.IsError() {
 		// Print the response status and error information.
 		e, _ := b["error"].(map[string]interface{})
-		return "", fmt.Errorf("[%s] type: %v, reason: %v", res.Status(), e["type"], e["reason"])
+		return fmt.Errorf("[%s] type: %v, reason: %v", res.Status(), e["type"], e["reason"])
 	}
 
 	// get the major version
 	version, _ := b["version"].(map[string]interface{})
 	number, _ := version["number"].(string)
 	if number == "" {
-		return "", fmt.Errorf("failed to detect elastic version number")
+		return fmt.Errorf("failed to detect elastic version number")
 	}
 
+	var c client
 	v := strings.Split(number, ".")[0]
-	return v, nil
+	switch v {
+	case ElasticV5:
+		c, err = v5.New(es.host, es.index)
+	case ElasticV6:
+		c, err = v6.New(es.host, es.index)
+	case ElasticV7:
+		c, err = v7.New(es.host, es.index)
+	default:
+		return fmt.Errorf("unsupported elasticsearch version %s", version)
+	}
+
+	es.mux.Lock()
```

Comment:
Why it only lock the assignment statement? This will cause multiple goroutines to initialize es client at the same time.

Reply:
true